### PR TITLE
Not real AWS mock version

### PR DESCRIPTION
**Pull request by vblaskovics**
2 strategies: stub, cloudsearch
2 search services: uploadDocuments, search

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (gergoerdosi) [inline .eslintrc]: `.eslintrc` is deprecated, use `.eslintrc.json` as the name of the file. Rename this in other projects too if you see it somewhere.

File: .gitignore [hunk 2]
```diff
@@ -31,3 +31,6 @@ node_modules
 
 # Optional REPL history
 .node_repl_history
+
+# Personal
+.idea
```
Comment (gergoerdosi) [inline .gitignore]: As it says, it's personal. It shouldn't be here. Use a global gitignore on your local environment instead. See the guide for more information: https://github.com/ideas/guides/tree/master/git#repository

File: config/test.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+'use strict';
```
Comment (gergoerdosi) [inline config/test.js]: I think this file can be removed, at least that's what we discussed.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@ideapod/hapi-aws-sqs",
+  "version": "0.0.1",
```
Comment (gergoerdosi) [inline package.json]: Use `0.0.0` as the initial version. Version number gets changed as part of the release process.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "@ideapod/hapi-aws-sqs",
```
Comment (gergoerdosi) [inline package.json]: Change the name to `@ideapod/hapi-aws-cloudsearch`.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  "version": "0.0.1",
+  "description": "AWS SQS integration plugin for hapi.",
```
Comment (gergoerdosi) [inline package.json]: Change the description to "AWS CloudSearch ..."

File: package.json [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  "main": "src/index.js",
+  "scripts": {
+    "test": "lab --assert code --lint --threshold 100 --verbose"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/ideas/hapi-aws-sqs.git"
```
Comment (gergoerdosi) [inline package.json]: Update the name of the module here and below too.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/ideas/hapi-aws-sqs/issues"
+  },
+  "homepage": "https://github.com/ideas/hapi-aws-sqs#readme",
+  "dependencies": {
+    "aws-sdk": "2.2.x",
+    "config": "^1.19.0",
+    "joi": "7.1.x",
+    "node-uuid": "1.4.x",
```
Comment (gergoerdosi) [inline package.json]: This doesn't seem to be used in this module.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    "twitter-text": "^1.13.3"
```
Comment (gergoerdosi) [inline package.json]: `twitter-text` is application specific, it shouldn't be in a general module.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  },
+  "devDependencies": {
+    "@ideapod/eslint-config": "ideas/eslint-config",
+    "code": "2.1.x",
+    "hapi": "12.0.x",
+    "lab": "8.0.x",
+    "mock-aws": "1.2.x",
+    "rewire": "^2.5.1"
```
Comment (gergoerdosi) [inline package.json]: As we discussed, use `sinon` if possible.

File: src/search.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+// Load external modules
```
Comment (gergoerdosi) [inline src/search.js]: Remove this comments if there aren't external modules to be loaded.

File: src/search.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+const CloudSearchStrategy = require('./strategies/cloudsearch');
+const StubStrategy = require('./strategies/stub');
+
+class Search {
+  constructor(options, simulate) {
+    this._settings = options;
+    this._strategy = simulate ? new StubStrategy(options) : new CloudSearchStrategy(options);
+  }
+
+  uploadDocuments(searchContent) {
```
Comment (gergoerdosi) [inline src/search.js]: If this is all the code does, it's enough to just use `return this._strategy..uploadDocument(this._settings, params)` instead of creating a new promise.

File: src/strategies/cloudsearch.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        return resolve(res);
+      });
+    });
+  }
+
+  addDocument(idea) {
+    // Remove certain UTF8 characters from the text
+    const utf8CharsRegex = /[^\u0009\u000a\u000d\u0020-\uD7FF\uE000-\uFFFD]/g;
+    const title = idea.title.replace(utf8CharsRegex, '');
```
Comment (gergoerdosi) [inline src/strategies/cloudsearch.js]: `title` and `body` are application specific fields, they shouldn't be specified in this module. Same with `idea`.

File: src/strategies/cloudsearch.js [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    const body = idea.body.replace(utf8CharsRegex, '');
+
+    let hashtags = Twitter.extractHashtags(idea.body);
```
Comment (gergoerdosi) [inline src/strategies/cloudsearch.js]: Again, hashtags are application specific.

File: src/strategies/stub.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+class StubStrategy {
+  constructor(options) {
```
Comment (gergoerdosi) [inline src/strategies/stub.js]: In this case the constructor can be removed.

File: test/index.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+      server = new Hapi.Server();
+      plugin = {
+        register: HapiAwsCloudsearch,
+        options: {
+          region: 'us-west-1',
+          endpoint: 'https://www.ideapod.com'
+        }
+      };
+      done();
+    });
```
Comment (gergoerdosi) [inline test/index.js]: Use new lines between test cases to make the code more readable.

File: test/index.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+lab.describe('HapiAwsCloudsearch', () => {
+  lab.describe('register', () => {
+    lab.beforeEach((done) => {
+      server = new Hapi.Server();
+      plugin = {
+        register: HapiAwsCloudsearch,
+        options: {
+          region: 'us-west-1',
+          endpoint: 'https://www.ideapod.com'
```
Comment (gergoerdosi) [inline test/index.js]: Use an endpoint that looks similar to an existing AWS CloudSearch endpoint.

File: test/search.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+const lab = exports.lab = Lab.script();
+const expect = Lab.assertions.expect;
+
+let searchParams;
+let search;
+
+lab.describe('#Search class', () => {
+  lab.describe('constructor', () => {
+    lab.beforeEach((done) => {
+      searchParams = {
```
Comment (gergoerdosi) [inline test/search.js]: If this doesn't change, then there is no need to use `beforeEach`.

File: test/search.js [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        region: 'us-west-1',
+        endpoint: 'https://www.ideapod.com'
+      };
+      done();
+    });
+
+    lab.it('constructs a Search object from this class', (done) => {
+      search = new Search(searchParams);
```
Comment (gergoerdosi) [inline test/search.js:26]: This is local to the test case, it shouldn't assign anything to an outside variable. Same below.

File: test/strategies/cloudsearch.js [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+    region: 'us-west-1',
+    endpoint: 'https://www.ideapod.com'
+  });
+  done();
+});
+
+lab.describe('CloudSearchStrategy', () => {
+  lab.describe('uploadDocuments', () => {
+    lab.beforeEach((done) => {
+      MockAWS.mock('CloudSearchDomain', 'uploadDocuments', (params, call) => {
```
Comment (gergoerdosi) [inline test/strategies/cloudsearch.js]: If `call` refers to `callback`, use `callback` as the name`,`call` is misleading.

File: test/strategies/cloudsearch.js [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+      strategy.search()
+        .catch((err) => {
+          expect(err).to.be.object();
+          done();
+        });
+    });
+  });
+
+  lab.describe('addDocument', () => {
+    let testIdea;
```
Comment (gergoerdosi) [inline test/strategies/cloudsearch.js]: Remove these. It's very hard to follow what's happening in the tests if you keep assigning data to outside variables.

File: .eslintrc (hunk 1 above)
Comment (vblaskovics) [inline .eslintrc]: Ok. We should fix it in hapi-aws-sqs repo too.

File: .gitignore (hunk 2 above)
Comment (vblaskovics) [inline .gitignore]: done

File: config/test.js (hunk 3 above)
Comment (vblaskovics) [inline config/test.js]: removed

File: package.json (hunk 4 above)
Comment (vblaskovics) [inline package.json]: done

File: package.json (hunk 5 above)
Comment (vblaskovics) [inline package.json]: done

File: package.json (hunk 6 above)
Comment (vblaskovics) [inline package.json]: done

File: package.json (hunk 7 above)
Comment (vblaskovics) [inline package.json]: done

File: test/search.js (hunk 18 above)
Comment (vblaskovics) [inline test/search.js]: done

File: src/strategies/stub.js (hunk 15 above)
Comment (vblaskovics) [inline src/strategies/stub.js]: done

File: test/index.js (hunk 16 above)
Comment (vblaskovics) [inline test/index.js]: done + remove beforeEach

File: test/index.js (hunk 17 above)
Comment (vblaskovics) [inline test/index.js]: done
